packages/web/src/features/translation/TranslationView.tsx

```diff
@@ -32,34 +30,58 @@ export default function TranslationView() {
   >([]);
   const queryClient = useQueryClient();
   const glossMutation = useMutation({
-    mutationFn: (variables: { wordId: string; gloss: string }) =>
+    mutationFn: (variables: { wordId: string; gloss?: string }) =>
       apiClient.words.updateGloss(variables.wordId, language, variables.gloss),
-    onMutate: ({ wordId }) => {
+    onMutate: async ({ wordId, gloss }) => {
       const requestId = Math.floor(Math.random() * 1000000);
       setGlossRequests((requests) => [...requests, { wordId, requestId }]);
-      return { requestId };
+
+      const queryKey = ['verse-glosses', language, params.verseId];
+      await queryClient.cancelQueries({ queryKey });
+      const previousGlosses = queryClient.getQueryData(queryKey);
+      queryClient.setQueryData<GetVerseGlossesResponseBody>(queryKey, (old) => {
+        if (old) {
+          const glosses = old.data.slice();
+          const index = glosses.findIndex((g) => g.wordId === wordId);
+          if (index >= 0) {
+            const doc = glosses[index];
+            glosses.splice(index, 1, {
+              ...doc,
+              approvedGloss: gloss,
+              glosses: gloss
+                ? doc.glosses.includes(gloss)
+                  ? doc.glosses
+                  : [...doc.glosses, gloss]
+                : doc.glosses,
+            });
+            return {
+              data: glosses,
+            };
+          }
+        }
+        return old;
+      });
+
+      return { requestId, previousGlosses };
     },
-    onError: () => {
+    onError: (_, __, context) => {
+      queryClient.setQueryData(
+        ['verse-glosses', language, params.verseId],
+        context?.previousGlosses
+      );
+
       alert('Unknown error occurred.');
     },
-    onSuccess: (_, { wordId, gloss }, context) => {
+    onSettled: (_, __, ___, context) => {
+      queryClient.invalidateQueries({
+        queryKey: ['verse-glosses', language, params.verseId],
+      });
```

Comment:
When the mutation finishes, then we instruct react-query to refetch that query to make sure it is up to date